django/db/models/expressions.py

```diff
@@ -538,6 +547,200 @@ def get_group_by_cols(self):
+            default = None
+        elif not hasattr(default, 'resolve_expression'):
+            default = F(default)
+        self.default = default
+
+    def get_source_expressions(self):
+        source_expressions = [value for condition, value in self.cases]
+        if self.default is not None:
+            source_expressions.append(self.default)
+        return source_expressions
+
+    def set_source_expressions(self, exprs):
+        if self.default is not None:
+            self.default = exprs[-1]
+            exprs = exprs[:-1]
+        assert len(exprs) == len(self.cases)
+        self.cases = [(condition, new_value) for new_value, (condition, value) in zip(exprs, self.cases)]
+
+    def get_source_fields(self):
+        """We're only interested in the fields of the value expressions."""
+        value_source_expressions = [value for condition, value in self.cases]
+        if self.default is not None:
+            value_source_expressions.append(self.default)
+        return [e._output_field_or_none for e in value_source_expressions]
+
+    def _parse_cases(self, cases):
+        raise NotImplementedError("Subclasses must implement _parse_cases()")
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False):
+        c = self.copy()
+        c.is_summary = summarize
+        cases = []
+        for condition, value in c.cases:
+            if hasattr(condition, 'resolve_expression'):
+                condition = condition.resolve_expression(query, allow_joins, reuse, summarize)
+            value = value.resolve_expression(query, allow_joins, reuse, summarize)
+            cases.append((condition, value))
+        c.cases = cases
+        if c.default is not None:
+            c.default = c.default.resolve_expression(query, allow_joins, reuse, summarize)
+        return c
+
+    def predicate_sql(self, compiler, connection):
+        raise NotImplementedError("Subclasses must implement predicate_sql()")
+
+    def condition_sql(self, condition, compiler, connection):
+        raise NotImplementedError("Subclasses must implement condition_sql()")
+
+    def value_sql(self, value, compiler, connection):
+        return compiler.compile(value)
+
+    def as_sql(self, compiler, connection):
+        if not self.cases:
+            if self.default is not None:
+                return compiler.compile(self.default)
+            return 'NULL', ()
+        predicate_sql, predicate_params = self.predicate_sql(compiler, connection)
+        result = ['CASE']
+        result_params = []
+        if predicate_sql:
+            result.append(predicate_sql)
+        result_params.extend(predicate_params)
+        for condition, value in self.cases:
+            condition_sql, condition_params = self.condition_sql(condition, compiler, connection)
+            value_sql, value_params = self.value_sql(value, compiler, connection)
+            result.append('WHEN %s THEN %s' % (condition_sql, value_sql))
```

Comment:
Should conditions be their own expressions that know how to render themselves? It could be an undocumented helper class of Case expressions:

```
class CaseCondition(Func):
    function = None
    template = 'WHEN %(condition)s THEN %(value)s'
    ...
```

It might reduce the size of the Case classes, but it might become too disconnected. Don't change it just because I've raised the idea, but it's something to think about.